test/functional/shared/resources/gateway_test.go

```diff
@@ -463,17 +461,45 @@ func testGatewayAvailability(hostname, baseURL, path string, expectedStatusCode
 	req.Host = hostname
 
 	client := newTestHTTPClient(isHttps, hostname)
-	res, err := client.Do(req)
-	if err != nil {
-		return err
-	}
 
-	if res.StatusCode != expectedStatusCode {
-		return fmt.Errorf("expected status code %d, got %d", expectedStatusCode, res.StatusCode)
+	retries := 2
+	retryBackoff := 5 * time.Second
+	for i := 1; i <= retries; i++ {
+		res, err := client.Do(req)
+		if err == nil && res.StatusCode == expectedStatusCode {
+			// Got expected status code, return
+			return nil
+		}
+
+		// If we got an error, or the status code was not what we expected, log the error and retry
+		// Logging the request and response will help with debugging the issue
+
+		t.Logf("failed to make request to %s with error: %s", urlPath, err)
+		requestDump, err := httputil.DumpRequestOut(req, true)
+		if err != nil {
+			t.Logf("failed to dump request with error: %s", err)
+		}
+		t.Logf("request dump: %s", string(requestDump))
+
+		if res == nil {
+			t.Logf("response is nil")
+		}
+
+		if res != nil && res.StatusCode != expectedStatusCode {
+			t.Logf("expected status code %d, got %d", expectedStatusCode, res.StatusCode)
+			responseDump, err := httputil.DumpResponse(res, true)
```

Comment:
Wouldn't we also want to dump the request (which was successful) in this case so we'd have that in case it contains the cause or a clue?

Reply:
(which was successfully sent so the code path to log it would get skipped is what I'm reading here)

Reply:
I think it should still dump the request - since that code block executes before this one. My logic was:
return if success
dump the request (if possible)
dump the response (if possible)

Reply:
I re-read the code path and yes I think it would log the request too every time as err would always be populated at that point in our failure case.